server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -241,139 +239,131 @@ synchronized void startWatcher(ClusterState clusterState, boolean onStartup) {
+                        // are one potential cause of inconsistency here, since their handling by
+                        // the WatchService is platform dependent.
+                        settingsDirWatchKey = enableSettingsWatcher(settingsDirWatchKey, operatorSettingsDir);
+
+                        if (watchedFileChanged(path)) {
+                            try {
+                                processFileSettings(path).get();
+                            } catch (ExecutionException e) {
+                                logger.error("Error processing operator settings json file", e.getCause());
+                            }
+                        }
+                    } catch (IOException e) {
+                        logger.warn("encountered I/O error while watching file settings", e);
                     }
+                } else {
+                    key.pollEvents();
+                    key.reset();
                 }
-            } catch (ClosedWatchServiceException | InterruptedException expected) {
-                logger.info("shutting down watcher thread");
-            } catch (Exception e) {
-                logger.error("shutting down watcher thread with exception", e);
-            } finally {
-                watcherThreadLatch.countDown();
             }
-        }, "elasticsearch[file-settings-watcher]").start();
+        } catch (ClosedWatchServiceException | InterruptedException expected) {
+            logger.info("shutting down watcher thread");
+        } catch (Exception e) {
+            logger.error("shutting down watcher thread with exception", e);
+        }
     }
 
     synchronized void stopWatcher() {
         logger.debug("stopping watcher ...");
         if (watching()) {
-            try {
-                // make sure the watcher thread hits the processing latch correctly
+            try (var ws = watchService) {   // make sure watch service is closed whatever
+                watcherThread.interrupt();
+                watcherThread.join();
+
+                // clean up the watch service
                 cleanupWatchKeys();
                 fileUpdateState = null;
-                watchService.close();
-                if (processingLatch != null) {
-                    processingLatch.countDown();
-                }
-                if (watcherThreadLatch != null) {
-                    watcherThreadLatch.await();
-                }
-                // the watcher thread might have snuck in behind us and re-created the settings watch again
-                cleanupWatchKeys();
             } catch (IOException e) {
                 logger.warn("encountered exception while closing watch service", e);
             } catch (InterruptedException interruptedException) {
-                logger.info("interrupted while closing the watch service", interruptedException);
+                logger.warn("interrupted while closing the watch service", interruptedException);
```

Comment:
This is interrupting the watch service, not the watch thread. What's the situations where the thread running the shutdown code will get interrupted?

Reply:
It can get interrupted by the node shutdown if it's taking too long, but also, let me ask a different question, what should be the action of an end user of our logs seeing a WARN message on interruption? If it's I/O exception, then it signals they are doing something strange with the file system that we rely on for the watcher, but what if the thread was interrupted?

Reply:
Ah, I didn't realise there was a valid situation for this code to be interrupted.